neuralhydrology/training/basetrainer.py

```diff
@@ -198,11 +199,19 @@ def train_and_validate(self):
         ``validate_every`` epochs. Model and optimizer state are saved after every ``save_weights_every`` epochs.
         """
         for epoch in range(self._epoch + 1, self._epoch + self.cfg.epochs + 1):
-            # set new learning rate
-            if epoch in self.cfg.learning_rate.keys():
-                LOGGER.info(f"Setting learning rate to {self.cfg.learning_rate[epoch]}")
-                for param_group in self.optimizer.param_groups:
-                    param_group["lr"] = self.cfg.learning_rate[epoch]
+            # Set new learning rate.
+            if self.cfg.learning_rate_drop_factor:
+                old_learning_rate = self._step_decay(epoch-1)
+                new_learning_rate = self._step_decay(epoch)
+                if old_learning_rate != new_learning_rate:
+                    LOGGER.info(f"Setting learning rate to {learning_rate}")
+                    for param_group in self.optimizer.param_groups:
+                        param_group["lr"] = self.learning_rate
```

Comment:
I am not super happy with that solution. I know that you want something that is more suited for hyper parameter tuning but personally I think that is a bit rushed.

If we want to add any other scheduler than what we had we should probably simply use the [learning rate schedulers](https://pytorch.org/docs/stable/optim.html#how-to-adjust-learning-rate) that are already implemented in PyTorch. E.g. what you implemented is `torch.optim.lr_scheduler.StepLR`, if I see it correctly. We could, similarly to everything else, have a `get_lr_scheduler()` function somewhere, where we could more easily integrate other schedulers in the future.

However, I think this should not be added in this PR (which is about the transformer) and in fact, I think all this learning rate scheduling should be removed for now. We could open a Issue here and discuss together how to add a more flexible way of learning rate scheduling.